design/variadic_config_alt.md

```diff
@@ -0,0 +1,637 @@
+apiVersion: kots.io/v1beta1 
+kind: ConfigValues 
+metadata: 
+  creationTimestamp: null 
+  name: qa-kots 
+spec: 
+  values: 
+    # EXISTING BEHAVIOR
+    <name>:
+      default: <value>
+      value: <value> 
+    # Example
+    a_templated_text: 
+      default: h6IVctWRdVBhflnQkImZybQUUkBjHQuAHj9QWFfBnFEOrf2CqBlkc70F22lMNHug 
+      value: GPyocL_6XLb4uCcvPhmoYKtnlWMX3mIHzopzUediHzRs1SenEpmJQi6fJqHDV6MX 
+    ...
+    # NEW!
+    # Values for a template
+    <prefix>-0-<name>:
+      parent: <templateGroup.Name>    # Not sure if this is needed, but seems like useful information, plus disambiguates from values that accidentally use the same syntax.
+      default: <value>
+      value: <value> 
+    # Example
+    kafka-0-hostname:
+      parent: kafka
+      value: kafka-zero.corp.com
+```
+
+```yaml
+apiVersion: kots.io/v1beta1 
+kind: ConfigValues 
+metadata: 
+  creationTimestamp: null 
+  name: qa-kots 
+spec: 
+  values: 
+    # EXISTING BEHAVIOR
+    <name>:
+      default: <value>
+      value: <value> 
+    # Example
+    a_templated_text: 
+      default: h6IVctWRdVBhflnQkImZybQUUkBjHQuAHj9QWFfBnFEOrf2CqBlkc70F22lMNHug 
+      value: GPyocL_6XLb4uCcvPhmoYKtnlWMX3mIHzopzUediHzRs1SenEpmJQi6fJqHDV6MX 
+    ...
+    # NEW!
+    # Values for a template
+    <prefix>-0-<name>:
+      parent: <templateGroup.Name>    # Not sure if this is needed, but seems like useful information, plus disambiguates from values that accidentally use the same syntax.
+      default: <value>
+      value: <value> 
+    # Example
+    kafka-name-0:
+      parent: kafka
+      groupId: 0
+      value: alpha
+    kafka-name-1:
+      parent: kafka
+      groupId: 1
+      value: bravo
+```
+
+### Resource Templates
+
+In addition to using the new syntax in the KOTS CRDS, they will template their resources using Golang Text Template `range` syntax like the following examples.
+Comments are used to keep valid syntax for linting purposes and also provide explicit documentation of generated fields.
+
+#### New ConfigContext Methods
+
+| Method            | Input | Output | Purpose |
+|-------------------|-------|--------|---------|
+| ConfigOptionList  |       |        |         |
+| ConfigOptionIndex |       |        |         |
+| ConfigGroupList   |       |        |         |
+|                   |       |        |         |
+|                   |       |        |         |
+
+#### Repeatable Config Item Usage
+
+Mounting a bunch of secrets (files) to a container as config data.
+
+```yaml
+# deployment.yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: my-deploy
+spec:
+  template:
+    spec:
+      containers:
+        - name: test
+          image: httpd
+          volumeMounts:
+            - mountPath: /var/www/html
+              name: secret-assets
+              readOnly: true
+      volumes:
+      - name: secret-assets
+        projected:
+          sources:
+          # repl{{ range $index, $value := ConfigOptionList "static_files" }}
```

Comment:
From an end user perspective we need to have a way to "select" what filename the static_file should be exposed as, for example file1: index.html, file2: favicon.ico ...